### ISSUE TITLE
[Feature]: Support The PQXDH Key Agreement Protocol

### ISSUE THREAD

**Issue by Yiannis128** [simplex-chat/simplex-chat#3097]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Platform

all

### Feature

Signal just released a white paper for the PQXDH Key Agreement Protocol, which is a quantum resistant upgrade to the Diffie-Hellman encryption. This issue suggests that Simplex should support such a protocol, since there is the rise of [quantum](https://www.datacenterdynamics.com/en/news/ibms-to-deploy-first-on-premise-quantum-computer-in-us/) [computers](https://spectrum.ieee.org/ibm-condor) that can defeat the standard encryption algorithms.

For reference: https://signal.org/docs/specifications/pqxdh/

**Comment by Yiannis128** [comment]
> There is only one problem: when PQXDH will be implemented several will not be able to exchange the key if one adopts the new agreement and the other the old one.

Perhaps could be done as an optional exchange method? So the current one is used as a fallback for those who don't accept.

**Comment by epoberezkin** [comment]
We are looking into augmenting current encryption primitives to become "believed to be resistant to quantum computer attacks".

We see is a priority, but not as an urgency.

It's important to remember that:
- there are no viable quantum computers right now, so we are talking about potentially mitigating "record now decrypt later" attacks.
- no existing algorithms are proven to be resistant, they are "believed to be resistant" (hence "potentially mitigating", rather than "mitigating").

A general problem with the standardised ML-KEM (aka CRYSTALS Kyber) algorithm:
- the step of hashing the output from the random number generator that was present in CRYSTALS-Kyber submission is removed from the algorithm, effectively making it vulnerable to the attack via a compromised RNG. See lines 304-314 in the published spec (https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.ipd.pdf) and also the linked discussion on the subject: https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/WFRDl8DqYQ4 . This decision polarised the experts, with some of them saying that it effectively created a backdoor.

Problems with PQXDH:
- it is "believed to" protect the initial key exchange from quantum computer attack, but it leaves authentication part unprotected (see part 4.1 of https://signal.org/docs/specifications/pqxdh/#authentication), even though there are existing quantum-resistant signature algorithms (and signature is used in this step anyway, so this part has no deniability).
- it does not PQ-protect break-in recovery of double ratchet (aka post compromise security) - it remains insecure to quantum computer attacks – protecting it would require the change in double ratchet itself.

Binding identity to key exchange and break-in recovery are very important qualities of double-ratchet design (without break-in recovery a very much simpler e2e encryption algorithm would suffice), and it's still work to do to protect them - PQXDH doesn't address it.

Problem with implementing PQXDH in SimpleX as is:
- encapsulation key for ML-KEM is ~1.5kb, so if it were added to the initial step of the handshake it would increase the invitation link size by 2kb (because of base 64 encoding), making it about 2500 characters (up from current 500). This is not viable without adding identity layer. It could be added in step 1 when the link is passed inside of the protocol (e.g. when connecting to group members).

The alternatives we are considering:
1) use ML-KEM, provided it is possible to mitigate the absence of hashing of random numbers. If it is possible, then we have two options:
  a) design and implement a modified version of double ratchet, that instead of adding ML-KEM to step 1 only, as PQXDH does, add it to step 2 of handshake and also to double ratchet key agreement steps that provide break-in recovery.
  b) wait until Signal adds post-quantum protection to double ratchet itself and implement that design - there are some non-trivial decisions to make, so waiting for Signal design could be wise.
 
Both a) and b) have the challenge that it would still take ~1500 bytes from the fixed size block, clashing with image and link previews, for example, so it requires some messaging protocol re-engineering that would support larger messages than fixed size block - this is possible, and probably this is what will happen before a/b decision is made.
2) use another algorithm, e.g. CTIDH - there are also some challenges with it though.

In any case, while PQXDH is "the first step in advancing quantum resistance for the Signal Protocol" (quoting the blog post), it is wrong to believe that it makes Signal Protocol "believed to be" quantum resistant – it does not yet.

We want to implement a bit more comprehensive design, and there is some work to do in preparation anyway, that we will be doing. Also, we may propose a modified version of double ratchet that adopts ML-KEM for break-in recovery steps for review.

**Comment by epoberezkin** [comment]
> Did you read the Tutanota approach?

not yet, will review later.

**Comment by epoberezkin** [comment]
There are two problems with the [modified double ratchet](https://eprint.iacr.org/2021/875.pdf) proposed by Tutanota:

1. They replaced DH in ratcheting steps with post-quantum KEM, rather than augmented it, therefore deviating from NIST recommendation to augment. This decision can make the scheme vulnerable to conventional computers, as resistance of post-quantum KEM to either quantum or conventional computers is not proven – it is "believed to be" resistant.
2. They added signature to each ratcheting step compromising deniability quality of double ratchet - now the sender cannot deny sending the first message in the sequence.

Both seem to introduce unnecessary limitations to what can be done by simply augmenting DH in each step with post-quantum KEM.

**Comment by epoberezkin** [comment]
Proposed modification to double ratchet: https://github.com/simplex-chat/simplex-chat/blob/ep/pqdr/docs/rfcs/2023-09-30-pq-double-ratchet.md (this scheme can work with any PQ KEM mechanism, it doesn't have to be Kyber).

Post by Daniel Bernstein about NIST standardisation process: https://blog.cr.yp.to/20231003-countcorrectly.html

**Comment by epoberezkin** [comment]
> Is it better than current one?

What is?

**Comment by epoberezkin** [comment]
@serrq yes, it's an improvement to the current implementation. Not implemented yet.

**Comment by agnosticlines** [comment]
Hey, just trying to understand, I see some parts of SimpleX are advertised as post quantum, such as the account/device linking? What is actually vulnerable to 'record now, decrypt later' currently?
